openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
+                    }
+                ))
+
+        # Write representations to the database
+        bulk_write(representation_writes)
+
+        # Backwards compatibility
+        # todo: can we avoid the need to store this?
+        instance.data["published_representations"] = {
+            p["representation"]["_id"]: p for p in prepared_representations
+        }
 
-        create_hard_link(src, dst)
+        self.log.info("Registered {} representations"
+                      "".format(len(prepared_representations)))
 
-    def get_subset(self, asset, instance):
+    def prepare_subset(self, instance):
+        asset = instance.data.get("assetEntity")
         subset_name = instance.data["subset"]
+        self.log.debug("Subset: {}".format(subset_name))
+
+        # Get existing subset if it exists
         subset = io.find_one({
             "type": "subset",
             "parent": asset["_id"],
             "name": subset_name
         })
 
-        if subset is None:
-            self.log.info("Subset '%s' not found, creating ..." % subset_name)
-            self.log.debug("families.  %s" % instance.data.get('families'))
-            self.log.debug(
-                "families.  %s" % type(instance.data.get('families')))
-
-            family = instance.data.get("family")
-            families = []
-            if family:
-                families.append(family)
+        # Define subset data
+        data = {
+            "families": get_instance_families(instance)
+        }
 
-            for _family in (instance.data.get("families") or []):
-                if _family not in families:
-                    families.append(_family)
+        subset_group = instance.data.get("subsetGroup")
+        if subset_group:
+            data["subsetGroup"] = subset_group
 
-            _id = io.insert_one({
+        bulk_writes = []
+        if subset is None:
+            # Create a new subset
+            self.log.info("Subset '%s' not found, creating ..." % subset_name)
+            subset = {
+                "_id": ObjectId(),
                 "schema": "openpype:subset-3.0",
                 "type": "subset",
                 "name": subset_name,
-                "data": {
-                    "families": families
-                },
+                "data": data,
                 "parent": asset["_id"]
-            }).inserted_id
+            }
+            bulk_writes.append(InsertOne(subset))
 
-            subset = io.find_one({"_id": _id})
+        else:
+            # Update existing subset data with new data and set in database.
+            # We also change the found subset in-place so we don't need to
+            # re-query the subset afterwards
+            subset["data"].update(data)
+            bulk_writes.append(UpdateOne(
+                {"type": "subset", "_id": subset["_id"]},
+                {"$set": {
+                    "data": subset["data"]
+                }}
+            ))
```

Comment:
Like before we currently update only the `data` of the existing Subset instead of replacing the document. Would it however be fine to just Replace the original document? Would there ever be any information on the existing Subset entity/document in the database that is outside of `data` which we want to preserve from the original subset?

If we could just replace the original document that would allow to simplify the code some more.

Reply:
>Subset entity/document in the database that is outside of data which we want to preserve from the original subset?

We should do all we can to preserve the ID in the mongo. It could be already referenced elsewhere. Even if not directly by our tool, then possibly by a studio custom tool relying on its persistence.